atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1404 @@
+    #[error("Failed to deserialize usage metrics: `{0}`")]
+    UsageMetricsDeserializeError(#[from] ciborium::de::Error<std::io::Error>),
+}
+
+mod utils {
+    use fastcrypto::{
+        ed25519::{Ed25519PublicKey, Ed25519Signature},
+        secp256k1::{Secp256k1PublicKey, Secp256k1Signature},
+        secp256r1::{Secp256r1PublicKey, Secp256r1Signature},
+        traits::{ToFromBytes as FastCryptoToFromBytes, VerifyingKey},
+    };
+    use flume::Sender;
+    use sui_sdk::types::{
+        base_types::SuiAddress,
+        crypto::{PublicKey, Signature, SignatureScheme, SuiSignature, ToFromBytes},
+    };
+    use tokio::sync::oneshot;
+    use tracing::{error, instrument};
+    use url::Url;
+
+    use crate::{types::UsageMetrics, AtomaP2pEvent};
+
+    use super::{AtomaP2pNodeError, StateManagerEvent};
+
+    /// The threshold for considering a timestamp as expired
+    const EXPIRED_TIMESTAMP_THRESHOLD: u64 = 10 * 60; // 10 minutes
+
+    /// Validates a UsageMetrics message by checking the URL format and timestamp freshness.
+    ///
+    /// This function performs two key validations:
+    /// 1. Ensures the node_public_url is a valid URL format
+    /// 2. Verifies the timestamp is within an acceptable time window relative to the current time
+    ///
+    /// # Arguments
+    ///
+    /// * `response` - The `UsageMetrics` containing the node_public_url and timestamp to validate
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` - If all validation checks pass
+    /// * `Err(AtomaP2pNodeError)` - If any validation fails
+    ///
+    /// # Errors
+    ///
+    /// Returns `AtomaP2pNodeError::InvalidPublicAddressError` when:
+    /// * The URL is invalid or malformed
+    /// * The timestamp is older than `EXPIRED_TIMESTAMP_THRESHOLD` (10 minutes)
+    /// * The timestamp is in the future
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// let metrics = UsageMetrics {
+    ///     node_public_url: "https://example.com".to_string(),
+    ///     timestamp: std::time::Instant::now().elapsed().as_secs(),
+    ///     signature: vec![],
+    ///     node_small_id: 1,
+    /// };
+    ///
+    /// match validate_public_address_message(&metrics) {
+    ///     Ok(()) => println!("UsageMetrics is valid"),
+    ///     Err(e) => println!("Invalid UsageMetrics: {}", e),
+    /// }
+    /// ```
+    ///
+    /// # Security Considerations
+    ///
+    /// This validation helps prevent:
+    /// * Invalid or malicious URLs from being propagated
+    /// * Replay attacks by enforcing timestamp freshness
+    /// * Future timestamp manipulation attempts
+    #[instrument(level = "debug", skip_all)]
+    pub fn validate_usage_metrics_url_and_timestamp(
+        usage_metrics: &UsageMetrics,
+    ) -> Result<(), AtomaP2pNodeError> {
+        let now = std::time::Instant::now().elapsed().as_secs();
+
+        // Check if the URL is valid
+        let _usage_metrics_url = Url::parse(&usage_metrics.node_public_url).map_err(|e| {
```

Comment:
I don't think it's sufficient to check if the url has a valid format, I think for now we should have an allow list of set urls providers who can join the mesh and report and remove any peers who have URLs outside that which try to subscribe. This can later be revisited as we define more rigid [peer scoring](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#peer-scoring)

Reply:
That's a great idea, I have created an issue for this. So that we can address it later

Reply:
See #390